packages/bundlers/default/src/DefaultBundler.js

```diff
@@ -1,191 +1,237 @@
+              entryAsset: asset,
               target: context.bundleGroup.target,
-              entryAssetId: asset.id
-            };
+              isEntry: context.bundleGroup.dependency.isEntry
+            });
+            bundleGraph.addBundleToBundleGroup(bundle, context.bundleGroup);
+            context.bundleByType.set(bundle.type, bundle);
+          }
+          bundleGraph.addAssetToBundle(asset, bundle);
+          bundleGraph.createAssetReference(dependency, asset);
+        }
+      }
+
+      return {
+        ...context,
+        parentNode: node
+      };
+    });
+  },
 
-            bundleGraph.addBundleGroup(context.bundle, bundleGroup);
-            context.bundleGroup = bundleGroup;
+  optimize({bundleGraph}) {
+    // Step 2: remove assets that are duplicated in a parent bundle
+    bundleGraph.traverseBundles({
+      exit(bundle) {
+        bundle.traverse((node, _, actions) => {
+          if (node.type !== 'dependency') {
+            return;
           }
 
-          if (
-            // this is the first bundle
-            !context.bundle ||
-            // this asset must be bundled on its own
-            asset.isIsolated ||
-            // the type of this asset differs from the current bundle type
-            asset.type !== context.bundle.type
-          ) {
-            // start by creating a new bundle
-            let bundle = assetGraph.createBundle(asset);
-            let dep = context.bundleGroup.dependency;
-
-            // Mark bundle as an entry, and set explicit file path from target if the dependency has one
-            bundle.isEntry = asset.isIsolated ? false : !!dep.isEntry;
-
-            if (!asset.isIsolated) {
-              // If the asset isn't isolated, try to combine with with an existing
-              // bundle.
-              let bundles = bundleGraph.getBundlesInBundleGroup(
-                context.bundleGroup
-              );
-              let existingBundle = bundles.find(b => b.type === asset.type);
-
-              // If there is an existing bundle of the asset's type, combine with that.
-              // Otherwise, a new bundle will be created.
-              if (existingBundle) {
-                existingBundle.merge(bundle);
-                return {
-                  bundleGroup: context.bundleGroup,
-                  bundle: existingBundle
-                };
-              }
-            }
+          let dependency = node.value;
+          let assets = bundleGraph.getDependencyAssets(dependency);
 
-            bundleGraph.addBundle(context.bundleGroup, bundle);
-            return {
-              bundleGroup: context.bundleGroup,
-              bundle
-            };
+          for (let asset of assets) {
+            if (bundleGraph.isAssetInAncestorBundles(bundle, asset)) {
+              bundleGraph.createAssetReference(dependency, asset);
+              bundleGraph.removeAssetTreeFromBundle(asset, bundle);
+              actions.skipChildren();
```

Comment:
Good catch. I probably missed this since often dependencies only have one asset. Not immediately sure how to approach this, but I'll explore and post here soon.

Reply:
We can probably get rid of this entirely. Since we're mutating the underlying graph when we remove the asset subgraph, we won't end up traversing downwards once that connection is gone.